### PR TITLE
Parallelize RING package tests

### LINKED ISSUE #3922: Slow tests
Some of our tests are slow. Here is list of packages with tests that run above 5 seconds (from https://github.com/cortexproject/cortex/runs/2054589567 run).

```
227.510s    github.com/cortexproject/cortex/pkg/ring
213.005s    github.com/cortexproject/cortex/pkg/querier
60.907s github.com/cortexproject/cortex/pkg/ingester
60.674s github.com/cortexproject/cortex/pkg/storegateway
42.779s github.com/cortexproject/cortex/pkg/chunk
39.780s github.com/cortexproject/cortex/pkg/chunk/encoding
39.727s github.com/cortexproject/cortex/pkg/ring/kv/memberlist
31.959s github.com/cortexproject/cortex/pkg/distributor
30.898s github.com/cortexproject/cortex/pkg/alertmanager
25.667s github.com/cortexproject/cortex/tools/blocksconvert/builder
23.770s github.com/cortexproject/cortex/pkg/chunk/purger
23.678s github.com/cortexproject/cortex/pkg/compactor
22.412s github.com/cortexproject/cortex/pkg/util
12.049s github.com/cortexproject/cortex/pkg/util/services
11.433s github.com/cortexproject/cortex/pkg/util/tls/test
8.013s  github.com/cortexproject/cortex/pkg/storage/tsdb/bucketindex
7.349s  github.com/cortexproject/cortex/pkg/chunk/storage
```

If we can optimize some of these, that would be great.

### PR DESCRIPTION
**Pull request by moki1202**
Signed-off-by: Shashank Singh [shashanksingh819@gmail.com](mailto:shashanksingh819@gmail.com)

**What this PR does**:
Optimizes go tests in `ring` package 

**Which issue(s) this PR fixes**:
Partially Fixes #3922 

Previous:
```
go test ./pkg/ring/ -v -count 1 -parallel 1
ok  	github.com/cortexproject/cortex/pkg/ring 71.475s
```
after:
```
go test ./pkg/ring/ -v -count 1
ok  	github.com/cortexproject/cortex/pkg/ring	10.188s

```
**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by moki1202** [comment]
@friedrichg tried my best, please let me know if this PR needs improvement. Wanted to ask a question, is it possible that using t.parallel() in some places might actually increase the already low test times rather than optimizing it further?

**Comment by friedrichg** [comment]
Thanks for your contribution.

@moki1202 As a general rule, it will use more cpus and will be faster. The only risk is we can make tests flaky if there is hidden bad concurrency in cortex.

Always sign your commits so it can pass the DCO. See [details](https://github.com/cortexproject/cortex/pull/5151/checks?check_run_id=11374077795)

**Comment by friedrichg** [comment]
Please look at the lint errors and fix them

**Comment by moki1202** [comment]
**On my machine**-
after-
```
go test -tags netgo -timeout 30m -race -count 1 ./pkg/ring/...
ok      github.com/cortexproject/cortex/pkg/ring        31.153s
```

**Review by friedrichg — CHANGES_REQUESTED** [review]
Awesome job! 

You are almost there, there is a few missing.

Following files still need some t.Parallel after t.Run

pkg/ring/kv/dynamodb/dynamodb_test.go
pkg/ring/kv/memberlist/memberlist_client_test.go
pkg/ring/model_test.go
pkg/ring/replication_set_tracker_test.go
pkg/ring/replication_strategy_test.go
pkg/ring/ring_test.go

**Comment by moki1202** [comment]
> Awesome job!
> 
> You are almost there, there is a few missing.
> 
> Following files still need some t.Parallel after t.Run
> 
> pkg/ring/kv/dynamodb/dynamodb_test.go pkg/ring/kv/memberlist/memberlist_client_test.go pkg/ring/model_test.go pkg/ring/replication_set_tracker_test.go pkg/ring/replication_strategy_test.go pkg/ring/ring_test.go

Done! Added t.parallel() to t.run() functions that I missed. Also added comment where parallelization wouldn't have worked.

**Review by friedrichg — APPROVED** [review]
Thanks for your patience. 

I noticed you skipped pkg/ring/kv/dynamodb/client_test.go too, but just add comments there, I don't think we can parallelize there too much.

Approved with minor nits.

**Comment by moki1202** [comment]
> Thanks for your patience.
> 
> I noticed you skipped pkg/ring/kv/dynamodb/client_test.go too, but just add comments there, I don't think we can parallelize there too much.
> 
> Approved with minor nits.

@friedrichg This is wierd, thought I'd covered every file. Will look into this again!

File: pkg/ring/model_test.go [hunk 1]
```diff
@@ -8,8 +8,7 @@ import (
 )
 
 func TestInstanceDesc_IsHealthy_ForIngesterOperations(t *testing.T) {
-	t.Parallel()
-
+	//parallel testing causes data race
```
[the remark this answers is not in the document]
Comment (moki1202) [reply pkg/ring/model_test.go:11]: @friedrichg Will check this again. Now I understand why you emphasized on creating smaller PRs 😅. Much easier to work with, that way.

**Pull request closed without merging**